File: src/main/java/com/assessment/user_service/service/UserService.java
package com.assessment.user_service.service;

import com.assessment.user_service.dto.CreateUserRequest;
import com.assessment.user_service.dto.UpdateUserRequest;
import com.assessment.user_service.exception.EmailAlreadyExistsException;
import com.assessment.user_service.exception.ResourceNotFoundException;
import com.assessment.user_service.model.User;
import com.assessment.user_service.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import java.util.Optional;

@Service
public class UserService {

    @Autowired
    private UserRepository userRepository;

    /**
     * Creates a new user.
     * @param request The request containing the user's details.
     * @return The created user.
     * @throws EmailAlreadyExistsException if the email already exists.
     */
    public User createUser(CreateUserRequest request) {
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new EmailAlreadyExistsException("An account with email '" + request.getEmail() + "' already exists.");
        }

        User user = new User();
        user.setName(request.getName());
        user.setEmail(request.getEmail());
        user.setAge(request.getAge());

        return userRepository.save(user);
    }

    /**
     * Returns a paginated list of all users.
     * @param pageable The pagination information.
     * @return A paginated list of users.
     */
    public Page<User> getAllUsers(Pageable pageable) {
        return userRepository.findAll(pageable);
    }

    /**
     * Returns a user by their ID.
     * @param id The ID of the user to retrieve.
     * @return An optional containing the user if found, or empty otherwise.
     */
    public Optional<User> getUserById(String id) {
        return userRepository.findById(id);
    }

    /**
     * Updates a user.
     * @param id The ID of the user to update.
     * @param userDetails The request containing the updated user details.
     * @return The updated user.
     * @throws ResourceNotFoundException if the user is not found.
     * @throws EmailAlreadyExistsException if the new email is already in use by another account.
     */
    public User updateUser(String id, UpdateUserRequest userDetails) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("User not found with id: " + id));

        if (userDetails.getName() != null) {
            user.setName(userDetails.getName());
        }
        if (userDetails.getAge() != null) {
            user.setAge(userDetails.getAge());
        }

        String newEmail = userDetails.getEmail();
        if (newEmail != null && !newEmail.equalsIgnoreCase(user.getEmail())) {
            if (userRepository.existsByEmailAndIdNot(newEmail, id)) {
                throw new EmailAlreadyExistsException("Email '" + newEmail + "' is already in use by another account.");
            }
            user.setEmail(newEmail);
        }

        return userRepository.save(user);
    }

    /**
     * Deletes a user.
     * @param id The ID of the user to delete.
     * @throws ResourceNotFoundException if the user is not found.
     */
    public void deleteUser(String id) {
        if (!userRepository.existsById(id)) {
            throw new ResourceNotFoundException("User not found with id: " + id);
        }
        userRepository.deleteById(id);
    }

}

File: src/services/userService.js
const User = require('../models/user.model');
const AppError = require('../utils/AppError');

// Creates a new user, ensuring the email is unique
const createUser = async (userData) => {
    const existingUser = await User.findOne({ email: userData.email });
    if (existingUser) {
        throw new AppError('An account with this email already exists.', 409);
    }

    const user = new User(userData);
    return await user.save();
};

// Fetches a paginated list of users, with an optional age filter
const getAllUsers = async (options) => {
    const { page, size, age } = options;

    const filterQuery = {};
    if (age) {
        filterQuery.age = age;
    }

    const users = await User.find(filterQuery)
        .limit(size)
        .skip(page * size)
        .select('-__v');

    return users;
};

// Fetches a single user by their ID
const getUserById = async (id) => {
    const user = await User.findById(id);
    if (!user) {
        throw new AppError(`No user found with ID: ${id}`, 404);
    }
    return user;
};

// Updates an existing user's information
const updateUser = async (id, updateData) => {
    if (updateData.email) {
        const emailExists = await User.findOne({
            email: updateData.email,
            _id: { $ne: id },
        });
        if (emailExists) {
            throw new AppError('This email is already in use by another account.', 409);
        }
    }

    const user = await User.findByIdAndUpdate(id, updateData, {
        new: true,
        runValidators: true,
    });

    if (!user) {
        throw new AppError(`No user found with ID: ${id} to update.`, 404);
    }

    return user;
};

// Deletes a user by their ID
const deleteUser = async (id) => {
    const user = await User.findByIdAndDelete(id);
    if (!user) {
        throw new AppError(`No user found with ID: ${id} to delete.`, 404);
    }
};

module.exports = {
    createUser,
    getAllUsers,
    getUserById,
    updateUser,
    deleteUser,
};